gazebo/rendering/LensFlare.hh

```diff
@@ -31,7 +32,80 @@ namespace gazebo
+
+      /// \brief Destructor
+      public: ~LensFlareCompositorListener();
+
+      /// \brief Set directional light that generates lens flare
+      /// \param[in] _light Pointer to directional light
+      public: void SetLight(LightPtr _light);
+
+      /// \brief Set the scale of lens flare.
+      /// \param[in] _scale Scale of lens flare
+      public: void SetScale(const double _scale);
+
+      /// \brief Set the color of lens flare.
+      /// \param[in] _color Color of lens flare
+      public: void SetColor(const ignition::math::Vector3d &_color);
+
+      /// \brief Callback that OGRE will invoke for us on each render call
+      /// \param[in] _passID OGRE material pass ID.
+      /// \param[in] _mat Pointer to OGRE material.
+      public: virtual void notifyMaterialRender(unsigned int _passId,
+                                                Ogre::MaterialPtr &_mat);
+
+      /// \brief Get the lens flare position and scale for a normal camera
+      /// \param[in] _camera Camera which the lens flare is added to
+      /// \param[out] _pos lens flare position in normalized device coordinates
+      /// \param[out] _scale Amount to scale the lens flare by.
+      private: void CameraPosScale(const CameraPtr &_camera,
+          ignition::math::Vector3d &_pos, double &_scale);
+
+      /// \brief Get the lens flare position and scale for a wide angle camera
+      /// \param[in] _wideAngleCam Camera which the lens flare is added to
+      /// \param[out] _pos lens flare position in normalized device coordinates
+      /// \param[out] _scale Amount to scale the lens flare by.
+      private: void WideAngleCameraPosScale(
+          const WideAngleCameraPtr &_wideAngleCam,
+          ignition::math::Vector3d &_pos, double &_scale);
+
+      /// \brief Check to see if the lens flare is occluded and return a scaling
+      /// factor that is proportional to the lens flare's visibility
+      /// \param[in] _cam Camera used for checking occlusion
+      /// \param[in] _imgPos light pos in clip space
+      /// \param[in] _worldPos light pos in 3D world space
+      private: double OcclusionScale(const CameraPtr &_cam,
+                                     const ignition::math::Vector3d &_imgPos,
+                                     const ignition::math::Vector3d &_worldPos);
+
+      /// \brief Pointer to light
+      private: LightPtr light;
+
+      /// \brief Pointer to camera
+      private: CameraPtr camera;
+
+      /// \brief Dummy camera used by wide angle camera for occlusion checking
+      private: CameraPtr wideAngleDummyCamera;
+
+      /// \brief Position of light in world frame
+      private: ignition::math::Vector3d lightWorldPos;
+
+      /// \brief Scale of lens flare.
+      private: double scale = 1.0;
+
+      /// \brief Color of lens flare.
+      private: ignition::math::Vector3d color
+          = ignition::math::Vector3d(1.0, 1.0, 1.0);
```

Comment:
if we are adding this class to the public API, then it should use the PIMPL pattern to allow more flexibility without breaking ABI

Reply:
Just to be sure that I am understanding your request correctly, you are suggesting to replace all the private member variables (i.e. `light` through `color`) with a `private: std::unique_ptr<LensFlareCompositorListenerPrivate> dataPtr`, where `LensFlareCompositorListenerPrivate` will be defined in the source file, just like how it is done with `LensFlare` and `LensFlarePrivate`?

Reply:
yes that is what I am proposing

I can assist with the modifications if needed

Reply:
thanks, on first look the PIMPL implementation looks great!

Reply:
FYI I'm still running some tests on my own end to verify it, just needed to push to test in a separate repo. One new addition is I need to init a LensFlareCompositorListenerPrivate in the constructor. I'll let you know when all my tests are working.

Reply:
New commit fixes it and passes my tests :)